man/dmgrc.1
.TH "DMGRC" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmgrc \(en creates log messages from Leica GeoCOM return codes
.SH SYNOPSIS
\f[B]dmgrc\f[R] \-\-help
.PP
\f[B]dmgrc\f[R] \-\-version
.PP
\f[B]dmgrc\f[R] [\-\-\f[B]node\f[R] \f[I]id\f[R]] [\-\-\f[B]level\f[R]
\f[I]level\f[R]] [\-\-\f[B]response\f[R] \f[I]name\f[R]]
[\-\-\f[B]name\f[R] \f[I]name\f[R]] [\-\-\f[B]logger\f[R]
\f[I]name\f[R]] [\-\-\f[B]debug\f[R]] [\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmgrc\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
The \f[B]dmgrc\f[R] program creates log messages from Leica GeoCOM
return codes.
Observations received by POSIX message queue are searched for a GeoCOM
return code (GRC) response.
If the code does not equal \f[CR]GRC_OK\f[R], a log message is sent to
the configured logger instance.
.PP
By default, observation responses of name \f[CR]grc\f[R] are verified.
For each GeoCOM error code, a custom log level may be specified in the
configuration file.
Otherwise, the default log level is used instead.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
File path to the optional configuration file.
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-level\f[R], \f[B]\-L\f[R] \f[I]level\f[R]
Default log level of return codes other than \f[CR]GRC_OK\f[R]
(\f[CR]LL_WARNING\f[R] by default).
The argument may be an integer or name string.
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of \f[I]dmlogger(1)\f[R] process to send logs to.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of program instance and configuration (\f[CR]dmgrc\f[R] by
default).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-response\f[R], \f[B]\-R\f[R] \f[I]name\f[R]
Response name of the GeoCOM return code (\f[CR]grc\f[R] by default).
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Process terminated without errors.
.TP
\f[B]1\f[R]
Failure.
Process failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
A configuration file is not required, but allows to specifiy the log
level of certain GeoCOM return codes.
In the following example configuration, the default log level for all
return codes other than \f[CR]GRC_OK\f[R] is set to
\f[CR]LL_WARNING\f[R].
The level is further refined for specific GeoCOM codes:
.IP
.EX
\-\- dmgrc.conf
dmgrc = {
  logger = \(dqdmlogger\(dq,
  node = \(dqdummy\-node\(dq,
  response = \(dqgrc\(dq,
  level = LL_WARNING,
  levels = {
    debug = { GRC_ABORT, GRC_SHUT_DOWN, GRC_NO_EVENT },
    info = { GRC_SLEEP_NODE, GRC_NA, GRC_STOPPED },
    warning = { GRC_TMC_ACCURACY_GUARANTEE, GRC_AUT_NO_TARGET },
    error = { GRC_FATAL },
    critical =
  },
  debug = false,
  verbose = true
}
.EE
.PP
Pass the path of the configuration file through the command\-line
argument:
.IP
.EX
$ dmgrc \-n dmgrc \-c /opt/etc/dmpack/dmgrc.conf
.EE
.PP
The name argument must match the name of the configuration table.


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` more error codes to `dm_error`
* `Added` more log levels to `dm_log`
* `Added` _fortran-zeromq_ interface bindings to `vendor/`
* `Added` function `dm_net_ipv6_is_valid()` to module `dm_net`
* `Added` type `posix_signal_type` and self-pipe procedures to module `dm_posix_signal`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_gantner` for eGateHighSpeedPort API access
* `Added` module `dm_group` for observation groups
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Added` module `dm_msgpack` for MessagePack serialisation/deserialisation
* `Added` module `dm_random` for pseudo-random number generation
* `Added` module `dm_zmq` for ZeroMQ access
* `Changed` location of sleep routines (moved to `dm_posix`)
* `Changed` name of module `dm_sql` to `dm_db_sql`
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_pipe`, `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` database schema due to new data model
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` GeoCOM API response identifiers to more memorable names
* `Changed` function `dm_path_join` in `dm_path`
* `Changed` model of observation data structure in `dm_observ`, removed priority, requests, and receivers
* `Changed` model of sensor data structure in `dm_sensor`
* `Deleted` module `dm_mqueue_util`
* `Deleted` module `dm_request`
* `Fixed` GeoCOM API response identifiers

## Programs

* `Changed` response value units in `dmved`
* `Changed` signal handlers to use self-pipe for POSIX conformance

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.1] – 2026-04-28

### Library

* `Added` example filter program to `share/dmfilter/`

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema

app/dmgrc.f90
! dmgrc.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmgrc
    !! Leica GeoCOM return code inspector. This program checks the GeoCOM return
    !! code (GRC) of observations from robotic total stations and creates a log
    !! message if an error is detected. The GRC response must be of type
    !! `RESPONSE_TYPE_INT32`.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmgrc'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    logical, parameter :: APP_MQ_BLOCKING = .true. !! Observation forwarding is blocking.

    type :: app_level_type
        integer, allocatable :: codes(:) !! Array of GeoCOM return codes.
    end type app_level_type

    type :: app_type
        !! Application settings.
        character(ID_LEN)            :: name      = APP_NAME   !! Instance and configuration name (required).
        character(FILE_PATH_LEN)     :: config    = ' '        !! Path to configuration file (required).
        character(LOGGER_NAME_LEN)   :: logger    = ' '        !! Name of logger.
        character(NODE_ID_LEN)       :: node_id   = ' '        !! Node id (required).
        character(RESPONSE_NAME_LEN) :: response  = 'grc'      !! Response name of GeoCOM return code.
        integer                      :: level     = LL_WARNING !! Default log level for return codes other than `GRC_OK`.
        logical                      :: debug     = .false.    !! Forward debug messages via IPC.
        logical                      :: verbose   = .false.    !! Print debug messages to stderr.
        type(app_level_type)         :: levels(LL_LAST)        !! Custom log levels of GeoCOM return codes.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer                 :: rc     ! Return code.
    type(app_type)          :: app    ! App settings.
    type(posix_mqueue_type) :: mqueue ! POSIX message queue.
    type(posix_signal_type) :: signal ! Self-pipe.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments, read settings from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Initialise logger.
    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward debug messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call logger%status('started ' // APP_NAME)

    rc = init(app, mqueue, signal)
    if (dm_is_error(rc)) call shutdown(rc)

    rc = run(app, mqueue, signal)
    call shutdown(rc)
contains
    integer function init(app, mqueue, signal) result(rc)
        !! Initialises program.
        type(app_type),          intent(in)  :: app    !! App type.
        type(posix_mqueue_type), intent(out) :: mqueue !! Message queue.
        type(posix_signal_type), intent(out) :: signal !! Self-pipe.

        ! Create self-pipe and register signal handler.
        signal_block: block
            rc = dm_posix_signal_create(signal);                            if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGINT,  signal_callback); if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGQUIT, signal_callback); if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGABRT, signal_callback); if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGTERM, signal_callback); if (dm_is_error(rc)) exit signal_block
        end block signal_block

        if (dm_is_error(rc)) then
            call logger%error('failed to initialize signal handler', error=rc)
            return
        end if

        call logger%debug('registered signal handler')

        ! Open observation message queue for reading.
        rc = dm_posix_mqueue_open(mqueue, type=TYPE_OBSERV, name=app%name, access=POSIX_MQUEUE_RDONLY)

        if (dm_is_error(rc)) then
            call dm_error_out(rc, 'failed to open mqueue /' // trim(app%name) // ': ' // dm_posix_error_message())
            return
        end if

        call logger%debug('opened mqueue /' // app%name)
    end function init

    subroutine find_level(levels, grc, level, default)
        !! Returns log level associated with GeoCOM return code in `level`.
        type(app_level_type), intent(in)  :: levels(:) !! Log levels of GeoCOM return codes.
        integer,              intent(in)  :: grc       !! GeoCOM return code to search for.
        integer,              intent(out) :: level     !! Associated log level or default.
        integer,              intent(in)  :: default   !! Default log level.

        integer :: i
        logical :: has

        level = default

        do i = 1, size(levels)
            has = dm_array_has(levels(i)%codes, grc)
            if (.not. has) cycle
            level = i
            return
        end do
    end subroutine find_level

    integer function run(app, mqueue, signal) result(rc)
        type(app_type),          intent(inout) :: app    !! App type.
        type(posix_mqueue_type), intent(inout) :: mqueue !! Message queue type.
        type(posix_signal_type), intent(inout) :: signal !! Self-pipe.

        integer           :: number
        logical           :: found
        type(observ_type) :: observ

        main_loop: do
            ! Poll for signals.
            rc = dm_posix_signal_poll(signal, timeout=0)

            if (rc == E_INTERRUPT) cycle main_loop

            if (dm_is_error(rc)) then
                call logger%error('failed to poll signal pipe', error=rc)
                exit main_loop
            end if

            if (dm_posix_signal_should_terminate(signal, number)) then
                call logger%debug('exit on signal ' // dm_posix_signal_name(number))
                exit main_loop
            end if

            ! Blocking read from POSIX message queue.
            call logger%debug('waiting for observation on mqueue /' // app%name)
            rc = dm_posix_mqueue_read(mqueue, observ)

            if (rc == E_INTERRUPT) cycle main_loop

            if (dm_is_error(rc)) then
                call logger%error('failed to read observation from mqueue /' // app%name, error=rc)
                call dm_posix_sleep(1)
                cycle main_loop
            end if

            ! Search all responses for GeoCOM return code.
            grc_block: block
                integer :: grc, grc_error, grc_type
                integer :: level

                found = .false.

                ! Find GeoCOM return code response.
                rc = dm_observ_get_response(observ, app%response, grc, type=grc_type, error=grc_error)
                if (dm_is_error(rc)) exit grc_block

                ! Validate response type.
                if (grc_type /= RESPONSE_TYPE_INT32) then
                    call logger%warning('GeoCOM return code response ' // trim(app%response) // ' of invalid type in observation ' // observ%name, observ=observ, error=E_TYPE)
                    exit grc_block
                end if

                ! Validate response error.
                if (dm_is_error(grc_error)) then
                    call logger%warning('invalid GeoCOM return code response ' // trim(app%response) // ' in observation ' // observ%name, observ=observ, error=grc_error)
                    exit grc_block
                end if

                ! Validate response value.
                if (dm_geocom_is_ok(grc)) then
                    call logger%debug('GeoCOM return code response ' // trim(app%response) // ' in observation ' // trim(observ%name) // ' not an error', observ=observ)
                    exit grc_block
                end if

                ! Find associated log level of GeoCOM return code, and create log message.
                call find_level(app%levels, grc, level, default=app%level)
                call logger%log(level, dm_geocom_error_message(grc), observ=observ, error=E_GEOCOM)

                found = .true.
            end block grc_block

            if (.not. found) call logger%debug('no GeoCOM return code response ' // trim(app%response) // ' found in observation ' // observ%name, observ=observ, error=E_NOT_FOUND)

            ! Forward observation.
            call logger%debug('finished observation ' // observ%name, observ=observ)
            rc = dm_posix_mqueue_forward(observ, name=app%name, blocking=APP_MQ_BLOCKING)
        end do main_loop
    end function run

    subroutine shutdown(error)
        !! Cleans up and stops program.
        integer, intent(in) :: error !! DMPACK error code.

        integer :: rc

        call dm_posix_signal_destroy(signal)

        call dm_posix_mqueue_close(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to close mqueue /' // app%name, error=rc)

        call dm_posix_mqueue_unlink(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to unlink mqueue /' // app%name, error=rc)

        call logger%status('stopped ' // APP_NAME, error=error)
        call dm_stop(merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error)))
    end subroutine shutdown

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app !! App type.

        integer                :: i
        type(arg_parser_class) :: parser

        call parser%add('name',     short='n', type=ARG_TYPE_ID)      ! -n, --name <id>
        call parser%add('config',   short='c', type=ARG_TYPE_FILE)    ! -c, --config <path>
        call parser%add('logger',   short='l', type=ARG_TYPE_ID)      ! -l, --logger <id>
        call parser%add('node',     short='N', type=ARG_TYPE_ID)      ! -N, --node <id>
        call parser%add('response', short='R', type=ARG_TYPE_ID, max_len=RESPONSE_NAME_LEN) ! -R, --response <id>
        call parser%add('level',    short='L', type=ARG_TYPE_LEVEL)   ! -L, --level <level>
        call parser%add('debug',    short='D', type=ARG_TYPE_LOGICAL) ! -D, --debug
        call parser%add('verbose',  short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Get all other arguments.
        call parser%get('logger',   app%logger)
        call parser%get('node',     app%node_id)
        call parser%get('response', app%response)
        call parser%get('level',    app%level)
        call parser%get('debug',    app%debug)
        call parser%get('verbose',  app%verbose)

        ! Allocate return code arrays.
        do i = 1, size(app%levels)
            if (.not. allocated(app%levels(i)%codes)) allocate (app%levels(i)%codes(0))
        end do

        ! Validate options.
        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name, geocom=.true.)

        if (dm_is_ok(rc)) then
            call config%get('logger',   app%logger)
            call config%get('node',     app%node_id)
            call config%get('response', app%response)
            call config%get('level',    app%level)
            call config%get('debug',    app%debug)
            call config%get('verbose',  app%verbose)

            if (dm_is_ok(config%field('codes'))) then
                call config%get('debug',    app%levels(LL_DEBUG   )%codes)
                call config%get('status',   app%levels(LL_STATUS  )%codes)
                call config%get('info',     app%levels(LL_INFO    )%codes)
                call config%get('warning',  app%levels(LL_WARNING )%codes)
                call config%get('error',    app%levels(LL_ERROR   )%codes)
                call config%get('critical', app%levels(LL_CRITICAL)%codes)
                call config%get('user1',    app%levels(LL_USER1   )%codes)
                call config%get('user2',    app%levels(LL_USER1   )%codes)

                call config%remove()
            end if
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(in) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_id_is_valid(app%name)) then
            call dm_error_out(rc, 'invalid name')
            return
        end if

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid node id')
            return
        end if

        if (.not. dm_id_is_valid(app%response, max_len=RESPONSE_NAME_LEN)) then
            call dm_error_out(rc, 'invalid response name')
            return
        end if

        if (.not. dm_log_level_is_valid(app%level)) then
            call dm_error_out(rc, 'invalid log level')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(number) bind(c)
        integer(c_int), intent(in), value :: number

        call dm_posix_signal_write(signal, number)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_lua_version(.true.)
    end subroutine version_callback
end program dmgrc
